zebrad/tests/common/get_block_template_rpcs/get_block_template.rs

```diff
@@ -112,7 +138,96 @@ pub(crate) async fn run() -> Result<()> {
+/// then calls getblocktemplate RPC in proposal mode with the serialized and hex-encoded data.
+///
+/// Returns an error if it fails to transform template to block proposal or serialize the block proposal
+/// Returns `Ok(())` if the block proposal is valid or an error with the reject-reason if the result is
+/// an `ErrorResponse`.
+///
+/// ## Panics
+///
+/// If an RPC call returns a failure
+/// If the response result cannot be deserialized to `GetBlockTemplate` in 'template' mode
+/// or `ProposalResponse` in 'proposal' mode.
+async fn try_validate_block_template(client: &RPCRequestClient) -> Result<()> {
+    let response_json_result = client
+        .json_result_from_call("getblocktemplate", "[]".to_string())
+        .await
+        .expect("response should be success output with with a serialized `GetBlockTemplate`");
+
+    tracing::info!(
+        ?response_json_result,
+        "got getblocktemplate response, hopefully with transactions"
+    );
+
+    // Propose a new block with an empty solution and nonce field
+    tracing::info!("calling getblocktemplate with a block proposal...",);
+
+    let raw_proposal_block =
+        hex::encode(proposal_block_from_template(response_json_result)?.zcash_serialize_to_vec()?);
+
+    let json_result = client
+        .json_result_from_call(
+            "getblocktemplate",
+            format!(r#"[{{"mode":"proposal","data":"{raw_proposal_block}"}}]"#),
+        )
+        .await
+        .expect("response should be success output with with a serialized `ProposalResponse`");
+
+    tracing::info!(?json_result, "got getblocktemplate proposal response");
+
+    match json_result {
+        ProposalResponse::Valid => Ok(()),
+        ProposalResponse::ErrorResponse { reject_reason, .. } => Err(eyre!(
+            "unsuccessful block proposal validation, reason: {reject_reason:?}"
+        )),
+    }
+}
+
+/// Make a block proposal from [`GetBlockTemplate`]
+fn proposal_block_from_template(
+    GetBlockTemplate {
+        version,
+        height,
+        previous_block_hash: GetBlockHash(previous_block_hash),
+        default_roots:
+            DefaultRoots {
+                merkle_root,
+                block_commitments_hash,
+                ..
+            },
+        bits: difficulty_threshold,
+        coinbase_txn,
+        transactions: tx_templates,
+        ..
+    }: GetBlockTemplate,
+) -> Result<Block> {
+    if Height(height) > Height::MAX {
+        Err(eyre!("height field must be lower than Height::MAX"))?;
+    };
+
+    let mut transactions = vec![coinbase_txn.data.as_ref().zcash_deserialize_into()?];
+
+    for tx_template in tx_templates {
+        transactions.push(tx_template.data.as_ref().zcash_deserialize_into()?);
+    }
 
-    Ok(())
+    Ok(Block {
+        header: Arc::new(block::Header {
+            version,
+            previous_block_hash,
+            merkle_root,
+            commitment_bytes: block_commitments_hash.into(),
+            time: Utc::now(),
```

Comment:
Using the raw local time could cause failures. The consensus rules require the `cur_time` from the block template, which has been clamped to the min and max times specified in the consensus rules.

Since we've had bugs in the time field, I'd like to test all 3 times here: `min_time`, `cur_time`, and `max_time`. All those times should produce valid blocks.

Reply:
Good point/idea, updated in [`516f17c`](https://github.com/ZcashFoundation/zebra/pull/5884/commits/516f17c257e3f39f63a9fa91ec0f0d253ca02b99) to return 3 block proposals from the template using `curtime`, `mintime`, and `maxtime`.